feat: resolve LB-type Service hostname to create A/AAAA instead of CNAME

cxuu opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

This PR adds an optional feature (turned off by default).

On EKS for example, the LoadBalancer-type Service object will not have external IP addresses but external hostnames. In this case, external-dns will always create an `Endpoint` (an external-dns concept, see its go [struct](https://pkg.go.dev/sigs.k8s.io/external-dns/endpoint#Endpoint)) whose `RecordType` is `CNAME`. But the user might want to create `A` or `AAAA` records instead, and perhaps on a different DNS provider (e.g. Cloudflare) than the cluster provider (e.g. AWS). 

This PR adds an optional feature that allows external-dns to resolve the LB hostname into IP addresses before creating the `Endpoint` struct.

<!-- Please provide a summary of the change here. -->

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #ISSUE

**Checklist**

- [x] Unit tests updated
- [x] End user documentation updated

Nefelim4ag commented:
I'm sorry, I can't seem to get what you are trying to solve here.
External-DNS can & will create CNAME on HostName load balancer, even if this is a different provider, from where Load Balancer Host it just will be CNAME.

On AWS much better solution is Aliases. About resolving - the only place where it "can" work as duct tape - with domain/zone name. Because we can not create CNAME Record `example.com` in Zone `example.com`.

Thanks!

reltuk commented:
FWIW, I think this PR is trying to solve the exact same issue as #2050.

reltuk commented:
> FWIW, I think this PR is trying to solve the exact same issue as #2050.

Never mind. I initially though the approach was related since I saw EKS, LoadBalancer-type Service and AAAA. While it may have been tangentially related, it doesn't seem to be solving the same problem.

benedikt-bartscher commented:
Hey @cxuu thanks for your PR. This would solve my problem with hetzner where i have to set the "load-balancer.hetzner.cloud/hostname: " annotation to fix some issues (see https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues/160). This causes external-dns to create CNAME records instead of A records which does not work:
`RRSet of type CNAME with DNS name ****.de. is not permitted at apex in zone ****.de.`

benedikt-bartscher commented:
@cxuu i just build an image for this PR and i found one bug. You named the argument "resolve-lb-hostname" in go but "resolve-load-balancer-hostname" in the helm chart.

cxuu commented:
@benedikt-bartscher Thank you for the bug report! I have fixed it in 1cf6530. I have also renamed the option to `resolveServiceLoadBalancerHostname` to indicate that it only works for Service. I imagine the code change for Ingress will look similar but personally I would prefer to add Ingress support in a second PR to make this PR smaller and reviewable.

szuecs commented:
/ok-to-test

stevehipwell commented:
@szuecs sorry I wasn't able to review this with you but I don't usually accept combined merges with code changes and Helm chart config.

@cxuu we don't accept flags as structure inputs unless they're required so I'm going to open a PR to remove the submitted Helm chart changes. Please use `extraArgs` for this purpose.

benedikt-bartscher commented:
I would also think about writing a seperate function which does the DNS Lookups. This function could then be used for both, Service and Ingress Resources. We could also implement a configureable timeout for the DNS Lookups.
